### PR TITLE
[Frontend] Add MiniMax H3 arbitrary timeline guide conditioning (GUIDE-01)

### PR DESCRIPTION
**Pull request by shykoe**
## Summary

Implements **GUIDE-01** from #7380: arbitrary timeline guide conditioning for
MiniMax H3, exposed through the `/v1/videos` and `/v1/videos/sync` HTTP API.

Guides are ordered conditioning anchors placed at explicit timeline positions.
Each guide carries an image or a video clip, optionally plus audio, and is
encoded through the VAE into the packed sequence without entering Qwen
presentation or consuming the ordinary reference budget.

Per the RFC task definition, this covers ordered image or clip guides, optional
audio, positive and negative frame indices, and valid `17k+5` guide lengths.

Closes GUIDE-01 of #7380.

## Motivation

`/v1/videos` already supports T2VA, first-frame FL2VA and mixed Ref2VA, but the
RFC baseline notes that *"arbitrary timeline guides ... are not yet exposed
through vLLM-Omni APIs."* Without them, the official ComfyUI Multiframe
Reference workflow cannot be executed remotely, which blocks the downstream
GUIDE-02 and WF-04 tasks.

### API surface

Two new multipart form fields on `POST /v1/videos` and `POST /v1/videos/sync`:

| Field | Type | Description |
|---|---|---|
| `timeline_guides` | JSON string | Ordered manifest. Each entry: `frame_index` (strict int) plus at least one of `image`, `video`, `audio`, each `{"upload_index": N}`. `image` and `video` cannot coexist in one entry. |
| `guide_files` | repeated upload | Binary files addressed by zero-based `upload_index`. Image JPEG/PNG/WebP, video MP4/MOV, audio WAV/MP3/FLAC. |

Guides require a guide-capable model, `quality=lossless` (cache-free execution),
and no active LoRA/Turbo adapter. Requests are rejected with 4xx when a model
does not declare the capability.

### Capability gating

`DiffusionModelMetadata.supports_timeline_guides` replaces hardcoded
pipeline-name checks, so `MiniMaxH3ModularPipeline` is accepted alongside
`MiniMaxH3Pipeline` instead of being silently rejected.

### Admission control

CPU-only validation runs before any GPU work: `TimelineGuideLimits`,
`TimelineGuideBudget`, and `validate_guide_descriptors` bound entry count,
unique file count, decoded pixels, and audio samples. Media decoding uses a
bounded `ffmpeg` subprocess with explicit frame and duration caps.

### ComfyUI-aligned placement

- `normalize_visual_frame_count` enforces the native `17k+5` cadence.
- `resolve_guide_start` resolves pixel-frame positions, including negative
  indices that address the tail of the timeline.
- `guide_audio_limit` bounds the audio window per guide.
- When guides are present, ordinary reference images use the ComfyUI `match`
  resize policy (area-preserving, down-only, 32-pixel aligned) rather than the
  2048-short-edge upscale.
- Per-condition noise uses an independent CPU RNG seeded from the request seed,
  so adding or removing one guide leaves other conditions bit-identical.

### Lifetime and error propagation

- `GuidedRequestLifetime` gives uploaded files cancellation-safe ownership;
  cleanup is gated on terminal worker provenance so an abort acknowledgment
  alone never authorizes deletion.
- Typed client errors raised inside GPU workers now survive the multiproc
  executor hop, preserving 4xx status instead of collapsing to a generic 500.
  Unknown or mixed fatal errors remain 500.

## Testing

Focused automated coverage, run from the repository root:

```bash
python -m pytest -q \
  tests/diffusion/models/minimax_h3/test_minimax_h3_timeline_guides.py \
  tests/diffusion/models/minimax_h3/test_minimax_h3_contract.py \
  tests/diffusion/models/minimax_h3/test_minimax_h3_packing.py \
  tests/diffusion/models/minimax_h3/test_minimax_h3_step_execution.py \
  tests/model_executor/stage_input_processors/test_minimax_h3.py
# 531 passed, 12 skipped

python -m pytest -q \
  tests/entrypoints/openai_api/test_video_server.py \
  tests/entrypoints/openai_api/test_video_api_utils.py \
  tests/entrypoints/openai_api/test_video_guided_lifetime.py
# 330 passed
```

Media probe/decode paths are exercised with real `ffmpeg`/`ffprobe` for
mono/stereo WAV, MP3 and FLAC, and for 12/30/60/600 FPS video including a
visible-pixel boundary case and a 22-frame tail clip.

## Real-model validation

Four H20 (96 GB) GPUs, FL2VA base weights, BF16, dense `FLASH_ATTN`, SP=4,
text-encoder TP=4, VAE patch parallelism=4, eager execution. No LoRA, cache,
quantization, or distilled schedule. All requests used 1344x768, 124 frames,
24 FPS, 50 steps, `quality=lossless`, seed 1101.

| Path | Result |
|---|---|
| Offline 22-frame tail clip at `-22` | 433.05 s |
| `/v1/videos/sync`, image + FLAC at frame 36 | HTTP 200, 726.63 s |
| `/v1/videos`, audio-only FLAC at frame 0, `preencode_mp4=true` | completed, 399.02 s, peak 66,174 MiB |
| `/v1/videos/sync`, same seed, no guides (control) | HTTP 200, 323.58 s |

All outputs passed full `ffmpeg -v error -i ... -f null -` decoding and contain
H.264 1344x768 / 124 frames / 24 FPS plus AAC 32 kHz stereo.

Conditioning effect versus the same-seed no-guide control:

- Image guide at frame 36: pixel MAE **4.65/255** vs **62.80/255** for the control.
- 22-frame tail clip: mean error **2.02/255** vs **85.68/255** for the control.
- 440 Hz audio guide: **99.85%** of spectral power in 430–450 Hz over its
  intended window, versus **0.0092%** for the same-prompt control.

Error handling was exercised live: six consecutive invalid guided sync requests
returned 400 without exhausting capacity, an invalid async placement produced a
structured failed job, DELETE removed it, a subsequent GET returned 404, and no
upload files remained on disk. A valid guide at `frame_index=-1` then returned
HTTP 200, confirming post-error recovery.

Full details are in `recipes/MiniMaxAI/MiniMax-H3.md`.

## Documentation and examples

- `docs/serving/videos_api.md` — request fields, placement and normalization
  rules, admission limits, cancellation semantics.
- `docs/design/feature/minimax_h3_timeline_guides.md` — design record.
- `recipes/MiniMaxAI/MiniMax-H3.md` — HTTP smoke command and validation
  record; the `Known limitations` section was also aligned with the supported
  reference matrix, as the RFC requests.
- `examples/online_serving/minimax_h3/` — server launch script, async and sync
  curl demos reproducing the official four-image reference case, and the shared
  prompt file they read.

No new Python example is added under `examples/`; the demo is shell-only and
drives the public HTTP API.
## Demo Result
correspondence with Demo in https://docs.comfy.org/tutorials/video/minimax/minimax-h3-multiframe

https://github.com/user-attachments/assets/6d981ca5-83c3-45ba-90ab-c2420d03a8d1

## Backward compatibility

No behavior changes for requests that omit `timeline_guides`. The new form
fields are optional, and the capability flag defaults to disabled for models
that do not declare it. The reserved `extra_args` key is rejected on public
ingress so it cannot be spoofed.

## Reviewer notes

1. **Duplicated reserved-key literal.** `"_minimax_h3_timeline_guides"` appears
   in both `vllm_omni/entrypoints/openai/protocol/videos.py` and
   `vllm_omni/entrypoints/openai/serving_video.py` to avoid importing the model
   module into the API layer. A comment explains the duplication, but a shared
   constant module would be cleaner if that import direction is acceptable.

2. **Offline validation provenance.** The recipe's real-model record includes
   two offline runs made with a local `Omni`-API driver script during
   validation. That script is not part of this PR; the recipe marks those
   entries accordingly, and the supported reproduction path is the HTTP smoke
   command plus the shell scripts under `examples/online_serving/minimax_h3/`.

Also noted for future cleanup, not addressed here: `validate_guide_descriptors`
and the HTTP admission check both enforce `max_entries`/`max_unique_files`
independently. The redundancy is harmless but could be consolidated.

## Out of scope

GUIDE-02 (ComfyUI extension support) and WF-04 (Multiframe Reference workflow
publication) depend on this task and are tracked separately in #7380.

### CODE REVIEW

**Comment by princepride** [comment]
@shykoe Hi, you need sign-off your commit first, pls refer: https://docs.vllm.ai/en/latest/contributing/#dco-and-signed-off-by

**Comment by shykoe** [comment]
> @shykoe Hi, you need sign-off your commit first, pls refer: https://docs.vllm.ai/en/latest/contributing/#dco-and-signed-off-by

I've added the sign-off to my commit and pushed the updated commit. Thanks for the reminder!

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py [hunk 1]
```diff
@@ -642,6 +651,38 @@ def _broadcast_rank0_exception(exc: Exception | None) -> None:
...
+            "error_type": getattr(exc, "error_type", None),
+        }
+    )
+    errors: list[Any] = [None] * world_size
+    dist.all_gather_object(errors, local_error, group=group)
+    for rank, info in enumerate(errors):
+        if info is None:
+            continue
+        message = f"[rank {rank}] {info['type']}: {info['message']}"
+        if info["status_code"] is not None:
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py]: Check every rank for fatal errors before returning a client error.

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py [hunk 2]
```diff
@@ -1701,6 +1725,205 @@ def _component_on_device(self, component: nn.Module):
...
+    @staticmethod
+    def _check_timeline_rows(limits: TimelineGuideLimits, guide_rows: int, other_rows: int) -> int:
+        packed_rows = ((guide_rows + other_rows + 63) // 64) * 64
+        if guide_rows > limits.max_guide_rows:
+            raise OmniClientError("MiniMax H3 timeline guides exceed the guide-row limit; trim guides")
+        if packed_rows > limits.max_packed_rows:
+            raise OmniClientError("MiniMax H3 timeline guides exceed the packed-request row limit; trim inputs")
+        return packed_rows
+
+    def _encode_timeline_guides(
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py]: Move timeline guide encoding and validation into a dedicated module.

File: examples/online_serving/minimax_h3/run_curl_minimax_h3_guides_sync.sh [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+EXTRA_PARAMS='{"task":"ref2va","aspect_ratio":"16:9","flow_shift":12.0,"audio_flow_shift":3.0}'
+
+echo "Generating MiniMax H3 Ref2VA + timeline guides (sync)..."
+echo "  Prompt file: ${PROMPT_FILE}"
+echo "  extra_params: ${EXTRA_PARAMS}"
+echo "This will block until generation completes."
+
+START_TIME=$(date +%s)
+
+curl -X POST "${BASE_URL}/v1/videos/sync" \
```
Comment (lishunyang12) [inline examples/online_serving/minimax_h3/run_curl_minimax_h3_guides_sync.sh]: Add --fail-with-body so HTTP errors don't get saved as successful videos.

**Review by FayeSpica — CHANGES_REQUESTED** [review]
Seems this semicolon breaks the Mermaid rendering. Could we use a comma?

File: docs/design/feature/minimax_h3_timeline_guides.md [hunk 4]
```diff
@@ -0,0 +1,636 @@
...
+    participant HTTP as HTTP caller
+    participant Bundle as Request bundle
+    participant Task as Inner task
+    participant Engine as Engine workers
+    participant Store as Output store
+    HTTP->>Bundle: reserve and persist inputs
+    Bundle->>Task: create strongly owned task
+    Task->>Engine: submit generation
+    HTTP-->>Bundle: timeout, cancellation or DELETE
+    Note over Bundle: set abandoned; keep inputs and capacity
```
Comment (FayeSpica) [inline docs/design/feature/minimax_h3_timeline_guides.md]:
```suggestion
    Note over Bundle: set abandoned, keep inputs and capacity
```

File: docs/design/feature/minimax_h3_timeline_guides.md [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+    Engine-->>Task: actual terminal completion
+    Task->>Bundle: check abandonment under publication lock
+    Note over Task,Store: discard output; do not recreate deleted job
```
Comment (FayeSpica) [inline docs/design/feature/minimax_h3_timeline_guides.md]:
```suggestion
    Note over Task,Store: discard output, do not recreate deleted job
```

**Review by FayeSpica — CHANGES_REQUESTED** [review]

**Comment by shykoe** [comment]
many conflicts 😭

**Comment by shykoe** [comment]
@vllm-omni-review-bot

**Comment by princepride** [comment]
@shykoe Seems we still have some conflicts, can you help resolve it, thank you.

**Comment by shykoe** [comment]
> @shykoe Seems we still have some conflicts, can you help resolve it, thank you.

resolved.

Self-Review 
Self-review for 1eeac643. Kilo assisted with the full branch review and
validation. The branch contains 53 changed files implementing MiniMax H3
arbitrary timeline-guide conditioning across the API, request lifetime,
distributed error transport, model preprocessing, packing, documentation,
and tests.

Core model path: guide placement, 17k+5 normalization, negative frame
indices, guide/reference row ordering, audio cropping, request/step parity,
no-guide compatibility, Cache-DiT restrictions, and distributed VAE
participation were reviewed. No blocking packing, frame, audio, or
parallelism defect was found.

API and lifetime path: strict multipart validation, capability gating,
bounded media admission, cancellation-safe file ownership, and multiprocess
typed-error transport are implemented. Two blocking gaps remain: uniproc
and non-step fallback paths discard typed 4xx metadata, and shutdown can
wait indefinitely for a guided task before engine teardown.

Validation completed:
- Model-focused suite: 553 passed, 14 skipped.
- API-focused suite: 369 passed.
- Combined focused suite: 1,194 passed, 14 skipped.
- Ruff, shellcheck, test marks, forbidden-import checks, torch.cuda policy,
  and git diff --check passed.

**Comment by linyueqian** [comment]
@shykoe this is 54 files and close to nine thousand added lines, and the branch currently conflicts with `main`. Please rebase, and if the guide conditioning can be split (for example the protocol and request schema, the MiniMax H3 model-side conditioning, and the serving and docs surface as separate PRs, or at least separate commits with their own tests), the review will be both faster and more useful than a single pass over the whole diff. I will pick it up as soon as a mergeable head is up.

**Comment by shykoe** [comment]
> @shykoe this is 54 files and close to nine thousand added lines, and the branch currently conflicts with `main`. Please rebase, and if the guide conditioning can be split (for example the protocol and request schema, the MiniMax H3 model-side conditioning, and the serving and docs surface as separate PRs, or at least separate commits with their own tests), the review will be both faster and more useful than a single pass over the whole diff. I will pick it up as soon as a mergeable head is up.

I'll split the changes into smaller, more manageable pieces to facilitate a faster and more useful review process.
